ape_safe/cli.py

```diff
@@ -0,0 +1,161 @@
+import inspect
+from typing import TYPE_CHECKING, Callable
+
+import click
+from ape.api.accounts import ImpersonatedAccount
+from ape.cli import (
+    ApeCliContextObject,
+    ConnectedProviderCommand,
+    account_option,
+    ape_cli_context,
+    network_option,
+)
+
+from ape_safe._cli.click_ext import safe_argument
+
+if TYPE_CHECKING:
+    from ape.api import AccountAPI, NetworkAPI
+
+    from ape_safe.accounts import SafeAccount
+
+
+def propose_from_simulation():
+    """
+    Create and propose a new SafeTx from transaction receipts inside a fork.
+
+    Usage::
+
+        @propose_from_simulation()
+        # NOTE: Name of decorated function *must* be called `cli`
+        # NOTE: Decorated function may have `safe` or `submitter` args in it
+        def cli():
+            # This entire function is executed within a fork/isolated context
+            # Use normal Ape features
+            my_contract = Contract("<address>")
+
+            # Any transaction is performed as if `sender=safe.address`
+            my_contract.mutableMethod(...)
+
+            # You can make assertions that will cause your simulation to fail if tripped
+            assert my_contract.viewMethod() == ...
+
+            # You can also add conditional calls
+            if my_contract.viewMethod() < some_number:
+                my_contract.mutableMethod()
+
+            # Once you are done with your transactions, the simulation will complete after exiting
+
+
+        # Once the simulation is complete, the decorator will collect all receipts it finds
+        # from the `safe` and collect them into a single SafeTx to propose to a public network
+    """
+
+    def decorator(
+        cmd: (
+            Callable[[], None]
+            | Callable[["SafeAccount"], None]
+            | Callable[["SafeAccount", "AccountAPI"], None]
+        ),
+    ) -> click.Command:
+        args: dict = {}
+        parameters = inspect.signature(cmd).parameters
+        if "safe" in parameters:
+            args["safe"] = None
+
+        if "submitter" in parameters:
+            args["submitter"] = None
+
+        if (script_name := cmd.__module__.split(".")[-1]).startswith("nonce"):
+            script_nonce = int(script_name[5:])
```

Comment:
this is kinda fragile and causes value errors if slightly named wrong like `something.nonce_7` and many other conditions.